addons/web/static/src/views/graph/graph_model.js

```diff
@@ -528,10 +523,17 @@ export class GraphModel extends Model {
      * @protected
      */
     async _prepareData() {
-        const processedDataPoints = this._getProcessedDataPoints();
+        let processedDataPoints = this._getProcessedDataPoints();
         this.data = null;
-        if (this._isValidData(processedDataPoints)) {
-            this.data = this._getData(processedDataPoints);
+        if (this._isValidData(processedDataPoints) && this.metaData.mode === 'pie') {
+            const positiveValues = [];
+            for (const dataPt of processedDataPoints) {
+                if (dataPt.value > 0) {
```

Comment:
Why isn't this processing in `_getProcessedDataPoints` ?
It's not clear to me what's the responsibility of `_prepareData` vs `_getProcessedDataPoints`
 
It also begs the question of the purpose of `_isValidData`. At the very least, its name is wrong.

Additional question: what does happen when all values are negative? does it work or does it display an empty pie chart?

Feel free to ignore this comment. I'm not part of @odoo/rd-framework-js and it's the first time I'm looking at this graph model.

Reply:
Hello @LucasLefevre 

> Why isn't this processing in `_getProcessedDataPoints` ? It's not clear to me what's the responsibility of `_prepareData` vs `_getProcessedDataPoints`

_getProcessedDataPoints filter/sort data points received from server
_prepareData creates from the processed dataPoint the data used by Chart.js

Node that the code chart.dataSource.getData() is in spreadsheet. The code in web was supposed to work with
this.data = null;

Now if you refactor the code as you do here, the separation will no more make sense of course.

> 
> It also begs the question of the purpose of `_isValidData`. At the very least, its name is wrong.

It does not make sense to have a pie chart that mixes negative and positive numbers. So from the point of view of a pie chart, the data is invalid (how to compute proportions?). Of course the data "in itself" is valid.

> 
> Additional question: what does happen when all values are negative? does it work or does it display an empty pie chart?

I think that Chart.js is able to present only negative numbers in a pie chart but it was decided (by NGR) that it would not be a nice feature. Hence the message displayed: 
Pie chart cannot mix positive and negative numbers. Try to change your domain to only display positive results

> 
> Feel free to ignore this comment. I'm not part of @odoo/rd-framework-js and it's the first time I'm looking at this graph model.

So you can rework the code with the new spec in head but I would remove _isValidData in that case and do stuff only in _getProcessedDataPoints.

Reply:
we'll have to simplify the template web.GraphView, ... and refactor the code in master.